jumanji/environments/swarms/common/updates.py

```diff
@@ -0,0 +1,183 @@
+    x: Tuple[chex.Array, types.AgentState],
+) -> Tuple[float, float]:
+    """
+    Get the updated agent heading and speeds from actions
+
+    Args:
+        _: Dummy JAX random key.
+        params: Agent parameters.
+        x: Agent rotation and acceleration actions.
+
+    Returns:
+        float: New agent heading.
+        float: New agent speed.
+    """
+    actions, boid = x
+    rotation = actions[0] * params.max_rotate * jnp.pi
+    acceleration = actions[1] * params.max_accelerate
+
+    new_heading = (boid.heading + rotation) % (2 * jnp.pi)
+    new_speeds = jnp.clip(
+        boid.speed + acceleration,
+        min=params.min_speed,
+        max=params.max_speed,
+    )
+
+    return new_heading, new_speeds
+
+
+@esquilax.transforms.amap
+def move(
+    _: chex.PRNGKey, _params: None, x: Tuple[chex.Array, float, float]
+) -> chex.Array:
+    """
+    Get updated agent positions from current speed and heading
+
+    Args:
+        _: Dummy JAX random key.
+        _params: unused parameters.
+        x: Tuple containing current agent position, heading and speed.
+
+    Returns:
+        jax array (float32): Updated agent position
+    """
+    pos, heading, speed = x
+    d_pos = jnp.array([speed * jnp.cos(heading), speed * jnp.sin(heading)])
+    return (pos + d_pos) % 1.0
+
+
+def init_state(
+    n: int, params: types.AgentParams, key: chex.PRNGKey
+) -> types.AgentState:
+    """
+    Randomly initialise state of a group of agents
+
+    Args:
+        n: Number of agents to initialise.
+        params: Agent parameters.
+        key: JAX random key.
+
+    Returns:
+        AgentState: Random agent states (i.e. position, headings, and speeds)
+    """
+    k1, k2, k3 = jax.random.split(key, 3)
+
+    positions = jax.random.uniform(k1, (n, 2))
+    speeds = jax.random.uniform(
+        k2, (n,), minval=params.min_speed, maxval=params.max_speed
+    )
+    headings = jax.random.uniform(k3, (n,), minval=0.0, maxval=2.0 * jax.numpy.pi)
+
+    return types.AgentState(
+        pos=positions,
+        speed=speeds,
+        heading=headings,
+    )
+
+
+def update_state(
+    key: chex.PRNGKey, params: AgentParams, state: types.AgentState, actions: chex.Array
+) -> types.AgentState:
+    """
+    Update the state of a group of agents from a sample of actions
+
+    Args:
+        key: Dummy JAX random key.
+        params: Agent parameters.
+        state: Current agent states.
+        actions: Agent actions, i.e. a 2D array of action for each agent.
+
+    Returns:
+        AgentState: Updated state of the agents after applying steering
+            actions and updating positions.
+    """
+    actions = jax.numpy.clip(actions, min=-1.0, max=1.0)
+    headings, speeds = update_velocity(key, params, (actions, state))
+    positions = move(key, None, (state.pos, headings, speeds))
+
+    return types.AgentState(
+        pos=positions,
+        speed=speeds,
+        heading=headings,
+    )
+
+
+def view(
+    _: chex.PRNGKey,
```

Comment:
Yeah Esquilax expects mapped functions to have a certain signature, though for cases like this it would be nice to be able to omit. I've not really settled on a clean way of having a function that I can optionally pass in arguments, but it is on y radar.
I'll add something in the docstring to explain.